src/qwed_new/phase4_runtime_probe.py

```
@@ -0,0 +1,56 @@
+def run_dynamic(user_expr):
+    return eval(user_expr)  # dynamic execution
+
+
+def run_shell(host):
+    os.system(f"ping -c1 {host}")  # shell execution
+    subprocess.call(f"nslookup {host}", shell=True)  # shell=True
+    subprocess.run(["dig", host])  # external process
+
+
+def deserialize(blob):
+    return pickle.loads(blob)  # unsafe deserialization
```

Ruff (0.15.21)

[error] `pickle` and modules that wrap it can be unsafe when used to deserialize untrusted data, possible security issue

(S301)